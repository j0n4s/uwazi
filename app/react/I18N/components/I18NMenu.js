import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { actions } from 'app/I18N';
import utils from '../utils';
import { NeedAuthorization } from 'app/Auth';
import { Icon } from 'UI';

export class I18NMenu extends Component {
  static reload(url) {
    window.location.href = url;
  }

  static saveCookie(locale) {
    utils.saveLocale(locale);
  }

  static changeLanguage(locale, url) {
    I18NMenu.saveCookie(locale);
    I18NMenu.reload(url);
  }

  constructor(props, context) {
    super(props, context);
    this.state = { open: false };
  }

  toggle() {
    this.setState({ open: !this.state.open });
  }

  inlineEditButton() {
    return (
      <NeedAuthorization roles={['admin', 'editor']}>
        <button
          className={this.props.i18nmode ? 'inlineEdit menuNav-btn btn btn-default active' : 'menuNav-btn btn btn-default'}
          onClick={this.props.toggleInlineEdit}
        >
          <Icon icon="language" size="lg" />
        </button>
      </NeedAuthorization>
    );
  }

  render() {
    const languages = this.props.languages.toJS();
    let path = this.props.location.pathname;
    const locale = utils.getLocale(path, languages);

    const regexp = new RegExp(`^\/?${locale}\/|^\/?${locale}$`);
    path = path.replace(regexp, '/');

    if (languages.length <= 1) {
      return (
        <ul className="menuNav-I18NMenu">
          {this.inlineEditButton()}
        </ul>
      );
    }

    return (
      <ul className="menuNav-I18NMenu">
        {this.inlineEditButton()}
        {(() => languages.map((lang) => {
            const url = `/${lang.key}${path}${path.match('document') ? '' : this.props.location.search}`;
            return (
              <li className={`menuNav-item${locale === lang.key ? ' is-active' : ''}`} key={lang.key}>
                <a
                  className="menuNav-btn btn btn-default"
                  href={url}
                  onClick={() => {
                  I18NMenu.changeLanguage(lang.key, url);
                }}
                >
                  {lang.key}
                </a>
              </li>
          );
          }))()}
      </ul>
    );
  }
}

I18NMenu.propTypes = {
  location: PropTypes.instanceOf(Object).isRequired,
  languages: PropTypes.instanceOf(Object).isRequired,
  toggleInlineEdit: PropTypes.func.isRequired,
  i18nmode: PropTypes.bool.isRequired
};

I18NMenu.contextTypes = {
  router: PropTypes.object
};

export function mapStateToProps(state) {
  return {
    languages: state.settings.collection.get('languages'),
    i18nmode: state.inlineEdit.get('inlineEdit')
  };
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators({ toggleInlineEdit: actions.toggleInlineEdit }, dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps)(I18NMenu);
