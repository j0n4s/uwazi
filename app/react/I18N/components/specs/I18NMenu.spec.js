import React from 'react';
import { shallow } from 'enzyme';
import Immutable from 'immutable';
import * as Cookie from 'tiny-cookie';
import { I18NMenu } from '../I18NMenu';

describe('I18NMenu', () => {
  let component;
  let props;

  beforeEach(() => {
    spyOn(Cookie, 'set');
    const languages = [
      { key: 'en', label: 'English', default: true },
      { key: 'es', label: 'EspaÃ±ol' }
    ];
    props = {
      languages: Immutable.fromJS(languages),
      toggleInlineEdit: jasmine.createSpy('toggleInlineEdit'),
      i18nmode: false,
      location: {
        pathname: '/templates/2452345',
        search: '?query=weneedmoreclerics'
      }
    };
  });

  const render = () => {
    component = shallow(<I18NMenu {...props} />);
    spyOn(I18NMenu, 'reload');
  };

  it('should not render searchQuery when on documents path', () => {
    props.location.pathname = '/es/documents';
    props.location.search = '?search';
    render();
    const links = component.find('a');
    expect(links.length).toBe(2);
    expect(links.first().props().href).toBe('/en/documents');
    expect(links.last().props().href).toBe('/es/documents');
  });

  describe('when there is NO language in the url', () => {
    it('should render links for each language', () => {
      render();
      const links = component.find('a');
      expect(links.length).toBe(2);
      expect(links.first().props().href).toBe('/en/templates/2452345?query=weneedmoreclerics');
      expect(links.last().props().href).toBe('/es/templates/2452345?query=weneedmoreclerics');
    });

    it('should work fine with triky urls', () => {
      props.location.pathname = '/entity/2452345';
      render();
      const links = component.find('a');
      expect(links.length).toBe(2);
      expect(links.first().props().href).toBe('/en/entity/2452345?query=weneedmoreclerics');
      expect(links.last().props().href).toBe('/es/entity/2452345?query=weneedmoreclerics');
    });

    it('should render as active the default language', () => {
      render();
      expect(component.find('li').first().props().className).toBe('menuNav-item is-active');
    });
  });

  describe('when there IS language in the url', () => {
    it('should render links for each language', () => {
      props.location.pathname = '/es/templates/2452345';
      render();
      const links = component.find('a');
      expect(links.length).toBe(2);
      expect(links.first().props().href).toBe('/en/templates/2452345?query=weneedmoreclerics');
      expect(links.last().props().href).toBe('/es/templates/2452345?query=weneedmoreclerics');
      expect(component.find('li').last().props().className).toBe('menuNav-item is-active');
    });

    it('should render as active the selected language', () => {
      props.location.pathname = '/es/templates/2452345';
      render();
      expect(component.find('li').last().props().className).toBe('menuNav-item is-active');
    });

    it('should render links for each language', () => {
      props.location.pathname = '/es';
      props.location.search = '';
      render();
      const links = component.find('a');
      expect(links.length).toBe(2);
      expect(links.first().props().href).toBe('/en/');
      expect(links.last().props().href).toBe('/es/');
    });
  });

  describe('when switching language', () => {
    it('should save the locale in to a coockie', () => {
      render();
      const links = component.find('a');
      links.first().simulate('click');
      expect(Cookie.set).toHaveBeenCalledWith('locale', 'en', { expires: 3650 });
    });
  });

  describe('when there is only one language', () => {
    it('should only render the inline translation button', () => {
      props.languages = Immutable.fromJS([{ key: 'en', label: 'English', default: true }]);
      render();
      expect(component).toMatchSnapshot();
    });
  });
});
