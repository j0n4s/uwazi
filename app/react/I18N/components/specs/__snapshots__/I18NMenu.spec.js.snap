// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`I18NMenu when there is only one language should only render the inline translation button 1`] = `
<ul
  className="menuNav-I18NMenu"
>
  <Connect(NeedAuthorization)
    roles={
      Array [
        "admin",
        "editor",
      ]
    }
  >
    <button
      className="menuNav-btn btn btn-default"
      onClick={[Function]}
    >
      <Component
        icon="language"
        size="lg"
      />
    </button>
  </Connect(NeedAuthorization)>
</ul>
`;
